torch/_refs/__init__.py

```diff
@@ -2825,7 +2825,13 @@ def masked_fill(a: TensorLikeType, mask: TensorLikeType, value: TensorOrNumberLi
             value_ndim == 0,
             lambda: f"only supports a 0-dimensional value tensor, but got tensor with {value_ndim} dimension",
         )
+        # `masked_fill`` allows cpu scalar to be moved to cuda but not otherwise.
+        check(
+            a.device.type == torch.device('cuda').type or value.device == a.device,
+            lambda: f"Expected `value` to be on same device as `a`."
+        )
         value_type = utils.dtype_to_type(value.dtype)
+        value = value.item()
```

Comment:
this forcibly introduces sync if value is on gpu, but `where` that's used in the decomp can handle gpu tensor fine (in fact, it will only handle gpu tensor)